crates/node-interaction/src/blocking_executor.rs

```diff
@@ -0,0 +1,205 @@
+//! The alloy crate __requires__ a tokio runtime.
+//! We contain any async rust right here.
+
+use std::{any::Any, panic::AssertUnwindSafe, pin::Pin, thread};
+
+use futures::FutureExt;
+use once_cell::sync::Lazy;
+use tokio::{
+    runtime::Builder,
+    sync::{mpsc::UnboundedSender, oneshot},
+};
+
+/// A blocking async executor.
+///
+/// This struct exposes the abstraction of a blocking async executor. It is a global and static
+/// executor which means that it doesn't require for new instances of it to be created, it's a
+/// singleton and can be accessed by any thread that wants to perform some async computation on the
+/// blocking executor thread.
+///
+/// The API of the blocking executor is created in a way so that it's very natural, simple to use,
+/// and unbounded to specific tasks or return types. The following is an example of using this
+/// executor to drive an async computation:
+///
+/// ```rust
+/// use revive_dt_node_interaction::*;
+///
+/// fn blocking_function() {
+///     let result = BlockingExecutor::execute(async move {
+///         tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+///         0xFFu8
+///     })
+///     .expect("Computation failed");
+///
+///     assert_eq!(result, 0xFF);
+/// }
+/// ```
+///
+/// Users get to pass in their async tasks without needing to worry about putting them in a [`Box`],
+/// [`Pin`], needing to perform down-casting, or the internal channel mechanism used by the runtime.
+/// To the user, it just looks like a function that converts some async code into sync code.
+///
+/// This struct also handled panics that occur in the passed futures and converts them into errors
+/// that can be handled by the user. This is done to allow the executor to be robust.
+///
+/// Internally, the executor communicates with the tokio runtime thread through channels which carry
+/// the [`TaskMessage`] and the results of the execution.
+pub struct BlockingExecutor;
+
+impl BlockingExecutor {
+    pub fn execute<R>(future: impl Future<Output = R> + Send + 'static) -> Result<R, anyhow::Error>
+    where
+        R: Send + 'static,
+    {
+        // A static of the state associated with the async runtime. This is initialized on the first
+        // access of the state.
+        static STATE: Lazy<ExecutorState> = Lazy::new(|| {
+            tracing::trace!("Initializing the BlockingExecutor state");
+
+            // Creating a multiple-producer-single-consumer channel which allows all of the other
+            // threads to communicate with this one async runtime thread.
+            let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<TaskMessage>();
```

Comment:
```suggestion
            let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<TaskMessage>();
```
Nit: Many of these comments are either overly obvious (superfluous) or would not be necessary with better variable namings. For example here just name the channel ends to something meaningful. Will also improve readability further down the line.

Ideally all code is expressive and readable enough to not need any comments. Comments should only be added to point out non-obvious and / or counterintuitive things / surprises. So if the code is readable and there are no unexpected things justifying a comment, all these comments just greatly disturb readability :)

I think your code is usually readable enough and don't need that many comments.